### PR TITLE
Time suite

### PR DESCRIPTION
**Pull request by eatkins**

### CODE REVIEW

**Comment by eatkins** [comment]
For whatever reason, I didn't see junk with zone.alloc, but I did with manual stdlib.malloc. At any rate, I added a helper to zero out the tm struct.

**Comment by eatkins** [comment]
I think it might make sense to do what is done for stat and create an alternative tm struct (in native c code, i.e. struct scalanative_tm) that only contains the platform independent fields: https://github.com/scala-native/scala-native/blob/master/nativelib/src/main/resources/stat.c

With that approach, we can expose only the standard portions of time.h (i.e. the fields that are in the existing struct defined in time.scala). If users want the extended functionality provided by their os, they can add their own native bindings.

Also, I am not at all committed to this PR. I just wanted an updated version of #1082 to pass against master. I only spent about two hours on it.

**Comment by eatkins** [comment]
Another approach to handle the size of the struct that should work is to define a time.c file and add
`size_t scalanative_sizeof_tm() { return sizeof(tm); }`
We could also add accessors
`int scalanative_tm_sec(tm *t) { return t->tm_sec; }`, etc.
The user api ends up being pretty kludgy though because you have to do things like
`val tmPtr: Ptr[tm] = stackalloc[scala.Byte](tm.getSize()).cast[Ptr[tm]]`
It makes more sense for things like pthread where there is no standard size for the opaque types like pthread_cond_t.

**Comment by densh** [comment]
Hi @eatkins, thanks for the contribution! Please run `bin/scalafmt` to unblock the CI.

**Comment by densh** [comment]
This PR also needs a rebase on top of latest master due to conflicting changes in e03a85c914bcda57f99647d0b3b9d0e671711831

**Comment by eatkins** [comment]
Whoops. Fixed everything up.

**Comment by eatkins** [comment]
Hmm... The ProcessSuite occasionally hangs in testing. My guess is that it's somehow related to signal handling. I'll have to look into that a bit further.

**Comment by densh** [comment]
@eatkins The failure looks spurious, we do get those on travis from time to time.

**Comment by Andrei-Pozolotin** [comment]
+1

File: nativelib/src/main/scala/scala/scalanative/native/time.scala [hunk 1]
```diff
@@ -1,44 +1,76 @@
...
+    def clock(): clock_t                                       = extern
+    def ctime_r(time: Ptr[time_t], buf: Ptr[CChar]): CString   = extern
+    def difftime(time_end: CLong, time_beg: CLong): CDouble    = extern
+    @name("scalanative_gmtime_r")
+    def gmtime_r(time: Ptr[time_t], tm: Ptr[tm]): Ptr[tm] = extern
+    @name("scalanative_localtime_r")
+    def localtime_r(time: Ptr[time_t], tm: Ptr[tm]): Ptr[tm] = extern
+    @name("scalanative_mktime")
+    def mktime(time: Ptr[tm]): time_t = extern
+    def strftime(str: Ptr[CChar],
```
Comment (Andrei-Pozolotin) [inline nativelib/src/main/scala/scala/scalanative/native/time.scala]: missing `@name("scalanative_strftime")` ?

Comment (eatkins) [reply nativelib/src/main/scala/scala/scalanative/native/time.scala]: Good catch. Should be fixed now.

File: nativelib/src/main/scala/scala/scalanative/native/time.scala [hunk 2]
```diff
@@ -1,44 +1,77 @@
...
-  def gmtime_s(time: Ptr[time_t], result: Ptr[tm]): Ptr[tm]    = extern
-  def localtime(time: Ptr[time_t]): Ptr[tm]                    = extern
-  def localtime_s(time: Ptr[time_t], result: Ptr[tm]): Ptr[tm] = extern
-  def mktime(time: Ptr[tm]): time_t                            = extern
+  private final val timeBufferSize = 26 // taken from `man asctime`
+  @inline private def zeroAlloc[T](size: CSize)(implicit z: Zone): Ptr[T] = {
+    val buf = z.alloc(size)
+    string.memset(buf, 0, size)
+    buf.cast[Ptr[T]]
+  }
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/time.scala]: In bindings we try to stay 100% close to the original signatures and APIs in C. We do not introduce extra parameters or helper methods. It's not the goal of bindings to increase usability of C code, it's the goal of the javalib or 3-rd party projects to build higher-level abstractions.

File: nativelib/src/main/scala/scala/scalanative/native/time.scala [hunk 3]
```diff
@@ -1,44 +1,77 @@
...
-  def tzname(): Ptr[CStruct2[CString, CString]] = extern
+               time: Ptr[tm]): CSize            = impl.strftime(str, count, format, time)
+  def time(arg: Ptr[time_t]): time_t            = impl.time(arg)
+  def tzset(): Unit                             = impl.tzset()
+  def daylight(): CInt                          = impl.daylight()
+  def timezone(): CLong                         = impl.timezone()
+  def tzname(): Ptr[CStruct2[CString, CString]] = impl.tzname()
+
+  @extern
+  private object impl {
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/time.scala]: `object time` itself has to be extern with all of its methods beings extern. As mentioned above bindings should stay true to the original C API, not provide an additional layer of abstraction.

File: nativelib/src/main/scala/scala/scalanative/native/time.scala (hunk 2 above)
Comment (eatkins) [reply nativelib/src/main/scala/scala/scalanative/native/time.scala]: Ok, if that's strictly the aim of the project, that's fine. I was just put off by the functions that returned naked pointers to static buffers.

File: nativelib/src/main/scala/scala/scalanative/native/time.scala (hunk 3 above)
Comment (eatkins) [reply nativelib/src/main/scala/scala/scalanative/native/time.scala]: Is there a technical reason for this or is it just the convention that scala native has adopted?

**Comment by eatkins** [comment]
Sort of related, should this time.scala file be moved to nativelib/src/main/scala/scala/scalanative/posix/ instead of nativelib/src/main/scala/scala/scalanative/native/?

**Comment by densh** [comment]
@eatkins Good point, it should be part of posix subpackage.

File: nativelib/src/main/scala/scala/scalanative/native/time.scala (hunk 3 above)
Comment (densh) [reply nativelib/src/main/scala/scala/scalanative/native/time.scala]: We need raw bindings and we also need higher-level abstractions. By separating the two we make it clear which is which. Some people may require access to lowest-level APIs without any abstractions, while others might prefer to have ergonomics-oriented higher-level version of it. Both need to exist.

Comment (eatkins) [reply nativelib/src/main/scala/scala/scalanative/native/time.scala]: That makes sense. Posix is what it is, for better and worse.

**Comment by eatkins** [comment]
I added a second commit that moves the file and cleans up some redundant declarations.

**Comment by densh** [comment]
LGTM, thanks for the contribution @eatkins ! 👍

**Pull request merged**